govcd/task.go

```diff
@@ -182,18 +192,253 @@ func (task *Task) GetTaskProgress() (string, error) {
+	if err != nil {
+		return nil, err
+	}
+	return client.GetTaskByHREF(taskHref)
+}
+
+// SkimTasksList checks a list of task IDs and returns  a list of IDs for tasks in progress and a list of IDs for failed ones
+func (client Client) SkimTasksList(taskIdList []string) ([]string, []string, error) {
+	var seenTasks = make(map[string]bool)
+	var newTaskList []string
+	var errorList []string
+	for i, taskId := range taskIdList {
+		_, seen := seenTasks[taskId]
+		if seen {
+			continue
+		}
+		seenTasks[taskId] = true
+		task, err := client.GetTaskById(taskId)
+		if err != nil {
+			if strings.Contains(err.Error(), errorRetrievingTask) {
+				// Task was not found. Probably expired. We don't need it anymore
+				continue
+			}
+			return newTaskList, errorList, err
+		}
+		util.Logger.Printf("[SkimTasksList] {%d} task %s %s (status %s - cancel requested: %v)\n", i, task.Task.Name, task.Task.ID, task.Task.Status, task.Task.CancelRequested)
+		if task.Task.Status == "success" || task.Task.Status == "aborted" {
+			continue
+		}
+		if task.Task.Status == "running" || task.Task.Status == "preRunning" || task.Task.Status == "queued" {
+			newTaskList = append(newTaskList, taskId)
+		}
+		if task.Task.Status == "error" && !task.Task.CancelRequested {
+			errorList = append(errorList, taskId)
+		}
+	}
+	return newTaskList, errorList, nil
+}
+
+// WaitTaskListCompletion waits until all tasks in the list are completed, removed, or failed
+// Returns a list of failed tasks and an error
+func (client Client) WaitTaskListCompletion(taskIdList []string, ignoreFailed bool) ([]string, error) {
+	var failedTaskList []string
+	var err error
+	for len(taskIdList) > 0 {
+		taskIdList, failedTaskList, err = client.SkimTasksList(taskIdList)
+		if err != nil {
+			return failedTaskList, err
+		}
+		time.Sleep(time.Second)
+	}
+	if len(failedTaskList) == 0 || ignoreFailed {
+		return nil, nil
+	}
+	return failedTaskList, fmt.Errorf("%d tasks have failed", len(failedTaskList))
+}
+
+// QueryTaskList performs a query for tasks according to a specific filter
+func (client *Client) QueryTaskList(filter map[string]string) ([]*types.QueryResultTaskRecordType, error) {
+	taskType := types.QtTask
+	if client.IsSysAdmin {
+		taskType = types.QtAdminTask
+	}
+
+	filterText := buildFilterText(filter)
+
+	notEncodedParams := map[string]string{
+		"type": taskType,
+	}
+	if filterText != "" {
+		notEncodedParams["filter"] = filterText
+	}
+	results, err := client.cumulativeQuery(taskType, nil, notEncodedParams)
+	if err != nil {
+		return nil, fmt.Errorf("error querying task %s", err)
+	}
+
+	if client.IsSysAdmin {
+		return results.Results.AdminTaskRecord, nil
+	} else {
+		return results.Results.TaskRecord, nil
+	}
+}
+
+// buildFilterText creates a filter with multiple values for a single column
+// Given a map entry "key": "value1,value2"
+// it creates a filter with a logical OR:  "key==value1,key==value2"
+func buildFilterText(filter map[string]string) string {
```

Comment:
Should this function be moved to `filter_util.go` (or similar file)?

Reply:
> Should this function be moved to filter_util.go (or similar file)?
So far, this kind of query is only used for Task.  
I can move it to `query.go`, but then we need to make sure that every query function understands this dialect, which is not standard.
I think we should improve the whole set of query functions, and add the ability of using logical OR, but it would require a lot of dedicated testing to do so. 
If I move this function to a general location, I may give the impression that this works for all query functions, which is not a given.

Reply:
Understood, thanks!